accepted/engine-host-functions.md

```diff
@@ -0,0 +1,211 @@
+// be used for satisfying imports
+let store = Store::new(module.engine());
+let linker = Linker::new(&store);
+ 
+let instance = linker.instantiate(&module)?;
+let run = instance
+    .get_export("run")
+    .unwrap()
+    .into_func()
+    .unwrap()
+    .get0::<()>();
+ 
+run()?;
+```
+ 
+## Changes to `wasmtime::Engine`
+ 
+This RFC proposes a `define_function` method be added to `Engine`:
+ 
+```rust
+/// Defines a function for the `Engine`.
+///
+/// Use `Store::get_engine_function` to get a `Func` representing the function.
+///
+/// A `Linker` will automatically default to functions on the associated `Engine`
+/// when resolving a function import that is not defined locally to the linker.
+pub fn define_function<Params, Results>(&self, module: &str, name: &str, func: impl IntoEngineFunc<Params, Results>) -> Result<()>;
+```
+
+Similar to `Func::wrap`, `define_function` will generically accept different `Fn` signatures to determine the WebAssembly type of the function. Instead of using `IntoFunc`, a different `IntoEngineFunc` trait will be used (see related section below).
+ 
+This method will internally create an `InstanceHandle` to represent the function.
+ 
+However, it will not associate the instance handle with a store; instead, the engine will own the associated instance handles and deallocate them when required (e.g. the `Engine` drops or an existing function definition is redefined).
+ 
+## Changes to `wasmtime::Store`
+ 
+To use `Engine` defined functions as imports for module instantiation or as `funcref` values, a `Func` representation is required.
+
+This proposal calls for adding the following methods to `Store`:
+ 
+```rust
+/// Gets a function from the `Engine` associated with this `Store`.
+///
+/// Returns `None` if the given function is not defined.
+pub fn get_engine_function(&self, module: &str, name: &str) -> Option<Func>;
+
+/// Get a reference to a context value of a particular type.
+///
+/// Returns `None` if the context value of the given type does not exist.
+pub fn get_context<T: Any>(&self) -> Option<&T>;
+ 
+/// Sets a context value.
+///
+/// Returns the previous context value of the same type if present.
+pub fn set_context<T: Any>(&mut self, value: T) -> Option<T>;
+ 
+/// Removes a context value.
+///
+/// Returns `None` if a context value of the given type does not exist.
+pub fn remove_context<T: Any>(&mut self) -> Option<T>;
+```
+ 
+`get_engine_function`  will register the function's instance handle with the `Store`, but as a *borrowed* handle which it is not responsible for deallocation. As `Store` keeps a reference on its associated `Engine`, this ownership model should be sufficient to prevent `funcref` values or imports of engine functions from outliving the function instance itself.
+
+For host functions that require contextual data (e.g. WASI), the `set_context` method will be used for associating context with a `Store` and it can later be retrieved via `caller.store().get_context()`.
```

Comment:
In use cases for API virtualization, the store doesn't seem like the right entity to attach contextual information to. Using the store here effectively makes all instances of an API share data.

The Wasmtime ABI passes the callee vmcontext to regular wasm functions; would it make sense to use a similar mechanism for engine functions, and store the state in something like the `host_state` field in `Instance`, instead of in the store?

Reply:
I'm not sure that instances are the best places to store data because they're generally designed at the wasm-spec-level to get decomposed into a bag of functions/globals/etc and those get passed around individually. The "calling instance" seems like it's sometimes a little ambiguous or just generally not available. 

For virtualization, though, can you detail how something like this wouldn't work well? Wasm-based virtualization wouldn't be affected at all since it's not storing data directly in the `Store`, and host-based virtualization presumably wouldn't be using the exact same type of context to virtualize? (and if it is it could always use a newttype)

Reply:
For more context, I originally proposed (in a privately reviewed draft) that the context be set on the instance, so I'm not opposed to that.

I think I was unclear in the intention of `set_context` allowing multiple values of different types, which is what Lucet does today with `insert_embed_ctx`.  As such, "insert" would probably be a better name as "set" has a singular connotation; I'll make that change.

If we do store it on the instance, we'll obviously need to modify the `wasmtime_runtime::Instance` to store the values and then `Caller` needs to be modified to perhaps return a `wasmtime::Instance` to get the values from?

That said, if we keep it on the `Store`, it's true that all instances using the `Wasi` host functions added at the engine-level will share the same `WasiCxt` context.  But that still does not preclude users from being able to call `Wasi::add_to_linker` to link in WASI with a different context for a specific instantiation.

Reply:
To be sure, I mean the *callee* vmctx arg, and not the *caller* one. The caller is indeed ambiguous and will ideally go away once interface types obviates exporting "memory". The callee, however, is pretty fundamental. Wasm functions need their instances.

The use case I'm thinking of is: an API which has some state associated with it. If two parts of an application each want to use that API, independently of the other, they should be able to have independent copies of the state. Ideally, we want host APIs to act as much like wasm APIs as possible, so that we can easily switch between host and polyfill implementations.

> That said, if we keep it on the Store, it's true that all instances using the Wasi host functions added at the engine-level will share the same WasiCxt context. But that still does not preclude users from being able to call Wasi::add_to_linker to link in WASI with a different context for a specific instantiation using the same store.

It's good that you can do multiple contexts; what I'm asking is, should this be the default way of doing things, and perhaps even the only way? Right now, there's not a lot attached to the store, and the state that is there is well understood, which is a useful property when we think about suspending/resuming/checkpoint-restart/migration/etc. if instances.

Here's a potentially crazy idea: each imported function has a `VMCallerCheckedAnyfunc`, which includes a `*mut VMContext`. But after instantiation and linking are done, one instantace doesn't ever look at the `*mut VMContext` for another instance. It just just blindly passes the context argument when calling an import. So, what if we generalized this pointer, and said it's no longer just a `*mut VMContext`, but an arbitrary "context" state pointer? For wasm modules, it'd still be the same `VMContext` pointer, but for host modules, it could be a pointer into whatever type the host wanted? Then we wouldn't even have to figure out how to fit this into `Instance`.

Reply:
Oh sorry I misunderstood about callee/caller. I'm curious though how you're thinking that the callee's state is initialized? For per-function state you can always simply capture state via the Rust closure, but the main problem with that is then the state is per-engine rather than per-instance/store. Where would per instance/store state get attached in what you're thinking @sunfishcode? (currently it's `set_context` for what this currently proposes I believe)

Reply:
Ok yeah, that is tricky. I'll have to think about this some more.

Reply:
@sunfishcode Do you have any suggestions as to an alternative approach? I'd like to resolve this prior to motion to finalize.